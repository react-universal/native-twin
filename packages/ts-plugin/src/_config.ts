export const pluginName = 'ts-styled-plugin-tw';
