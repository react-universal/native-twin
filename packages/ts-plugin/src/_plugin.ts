// @ts-nocheck
import {
  decorateWithTemplateLanguageService,
  TemplateSettings,
} from 'typescript-template-language-service-decorator';
import ts from 'typescript/lib/tsserverlibrary';
import { ConfigurationManager } from './_configuration';
import { StyledTemplateLanguageService } from './_language-service';
import { LanguageServiceLogger } from './_logger';
import { getSubstitutions } from './_substituter';
import { StyledVirtualDocumentFactory } from './_virtual-document-provider';

export class StyledPlugin {
  private _logger?: LanguageServiceLogger;
  private readonly _configManager = new ConfigurationManager();

  public constructor(private readonly typescript: typeof ts) {}

  public create(info: ts.server.PluginCreateInfo): ts.LanguageService {
    this._logger = new LanguageServiceLogger(info);
    this._configManager.updateFromPluginConfig(info.config);

    this._logger.log('config: ' + JSON.stringify(this._configManager.config));

    if (!isValidTypeScriptVersion(this.typescript)) {
      this._logger.log('Invalid typescript version detected. TypeScript 3.x required.');
      return info.languageService;
    }

    return decorateWithTemplateLanguageService(
      this.typescript,
      info.languageService,
      info.project,
      new StyledTemplateLanguageService(
        this.typescript,
        this._configManager,
        new StyledVirtualDocumentFactory(),
        this._logger,
      ),
      getTemplateSettings(this._configManager),
      { logger: this._logger },
    );
  }

  public onConfigurationChanged(config: any) {
    if (this._logger) {
      this._logger.log('onConfigurationChanged');
    }
    this._configManager.updateFromPluginConfig(config);
  }
}

export function getTemplateSettings(configManager: ConfigurationManager): TemplateSettings {
  return {
    get tags() {
      return configManager.config.tags;
    },
    enableForStringWithSubstitutions: true,
    getSubstitutions(templateString, spans): string {
      return getSubstitutions(templateString, spans);
    },
  };
}

function isValidTypeScriptVersion(typescript: typeof ts): boolean {
  const [major] = typescript.version.split('.');
  return +major! >= 3;
}
