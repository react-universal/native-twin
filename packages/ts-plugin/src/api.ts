export { StyledTemplateLanguageService } from './_language-service';
export { StyledPluginConfiguration } from './_configuration';
export { VirtualDocumentProvider } from './_virtual-document-provider';
export { getTemplateSettings } from './_plugin';
