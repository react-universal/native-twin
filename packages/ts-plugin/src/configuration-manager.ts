export interface StyledPluginConfiguration {
  readonly tags: ReadonlyArray<string>;
  readonly validate: boolean;
  readonly lint: { [key: string]: any };
  readonly emmet: { [key: string]: any };
}

export class ConfigurationManager {
  private static readonly defaultConfiguration: StyledPluginConfiguration = {
    tags: ['styled', 'css', 'extend', 'injectGlobal', 'createGlobalStyle', 'keyframes'],
    validate: true,
    lint: {
      emptyRules: 'ignore',
    },
    emmet: {},
  };

  private readonly _configUpdatedListeners = new Set<() => void>();

  public get config(): StyledPluginConfiguration {
    return this._configuration;
  }
  private _configuration: StyledPluginConfiguration =
    ConfigurationManager.defaultConfiguration;

  public updateFromPluginConfig(config: StyledPluginConfiguration) {
    const lint = Object.assign(
      {},
      ConfigurationManager.defaultConfiguration.lint,
      config.lint || {},
    );

    this._configuration = {
      tags: config.tags || ConfigurationManager.defaultConfiguration.tags,
      validate:
        typeof config.validate !== 'undefined'
          ? config.validate
          : ConfigurationManager.defaultConfiguration.validate,
      lint,
      emmet: config.emmet || ConfigurationManager.defaultConfiguration.emmet,
    };

    for (const listener of this._configUpdatedListeners) {
      listener();
    }
  }

  public onUpdatedConfig(listener: () => void) {
    this._configUpdatedListeners.add(listener);
  }
}
