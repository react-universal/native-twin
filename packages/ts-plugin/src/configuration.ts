import { TwindPluginConfiguration } from './types';

export class ConfigurationManager {
  private static readonly defaultConfiguration: TwindPluginConfiguration = {
    tags: ['tw', 'apply', 'css', 'styled', 'variants'],
    attributes: ['tw', 'class', 'className', 'variants'],
    styles: ['style', 'styled'],
    debug: false,
    enable: true,
  };

  private readonly _configUpdatedListeners = new Set<() => void>();

  public get config(): TwindPluginConfiguration {
    return this._configuration;
  }

  private _configuration: TwindPluginConfiguration = {
    ...ConfigurationManager.defaultConfiguration,
    tags: [...ConfigurationManager.defaultConfiguration.tags],
  };

  public updateFromPluginConfig(config: Partial<TwindPluginConfiguration> = {}): void {
    const mergedConfig = {
      ...ConfigurationManager.defaultConfiguration,
      ...config,
    };

    this._configuration = {
      ...mergedConfig,
      debug: 'true' == String(mergedConfig.debug),
      enable: 'false' != String(mergedConfig.enable),
    };

    for (const listener of this._configUpdatedListeners) {
      listener();
    }
  }

  public onUpdatedConfig(listener: () => void): void {
    this._configUpdatedListeners.add(listener);
  }
}
