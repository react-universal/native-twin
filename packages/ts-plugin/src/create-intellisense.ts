import type {
  BaseTheme,
  ExtractThemes,
  Preset,
  Twind,
  TwindConfig,
  TwindUserConfig,
} from '@universal-labs/twind-adapter';
import { matchSorter, type MatchSorterOptions } from 'match-sorter';
import cssbeautify from 'cssbeautify';
import { TailwindTheme } from '@twind/preset-tailwind';
import QuickLRU from 'quick-lru';
import type { CurrentTheme, Intellisense, IntellisenseOptions, Suggestion } from './types';
import { createIntellisenseContext } from './internal/create-context';
import { spacify } from './utils';
import { compareSuggestions } from './internal/compare-suggestion';
export * from './types';

export function createIntellisense<Theme extends BaseTheme = BaseTheme>(
  twind: Twind<Theme>,
  options?: IntellisenseOptions,
): Intellisense<Theme & TailwindTheme>;

export function createIntellisense<Theme extends BaseTheme = BaseTheme>(
  config: TwindConfig<Theme>,
  options?: IntellisenseOptions,
): Intellisense<Theme & TailwindTheme>;

export function createIntellisense<
  Theme = BaseTheme,
  Presets extends Preset<any>[] = Preset[],
>(
  config: TwindUserConfig<Theme, Presets>,
  options?: IntellisenseOptions,
): Intellisense<CurrentTheme & ExtractThemes<Theme, Presets>>;

export function createIntellisense(
  config: Twind | TwindConfig | TwindUserConfig,
  options: IntellisenseOptions = {},
): Intellisense {
  const suggestionCache = new QuickLRU<string, Suggestion[]>({
    maxSize: 1000,
    ...options.cache,
  });

  const context = createIntellisenseContext(config as Twind<CurrentTheme>, options);

  // Precache empty input as it is the most common and take a while
  suggestionCache.set('', context.suggestions.map(toSuggestion));

  return {
    get theme() {
      return context.tw.theme;
    },
    get config() {
      return context.tw.config;
    },
    async suggest(input, { prefix = '', ignore } = {}) {
      const key = JSON.stringify({ input, prefix, ignore });
      let result = suggestionCache.get(key);

      if (!result) {
        let source = context.suggestions;
        let threshold: MatchSorterOptions['threshold'] = matchSorter.rankings.MATCHES;

        const token = prefix + input;

        const match = /^(.+(?:-\[.+])?)\/([^/]+|\[.+])?$/.exec(token);
        if (match) {
          const { 1: key, 2: modifier = '' } = match;
          const suggestion = context.classes.get(key!) || context.variants.get(key!);
          if (suggestion?.modifiers) {
            source = suggestion?.modifiers;
            input = modifier;
          }
        }

        if (token[0] === '-') {
          source = source.filter(({ name }) => name[0] === '-');

          if (prefix) {
            prefix = prefix.slice(1);
          } else {
            input = input.slice(1);
          }
        }

        if (ignore?.length) {
          source = source.filter(({ value }) => !ignore.includes(value));
        }

        if (prefix) {
          source = source
            .filter(({ type, value }) => type === 'variant' || value.startsWith(prefix))
            .map((suggestion) =>
              suggestion.type === 'variant'
                ? suggestion
                : {
                    ...suggestion,
                    filter: spacify(suggestion.value.slice(prefix.length)),
                    value: suggestion.value.slice(prefix.length),
                  },
            );
        }

        const { length } = input.replace(/[-\s]+/g, '');

        if (length === 0) {
          if (prefix) {
            source = source.sort((a, b) => compareSuggestions(a, b, prefix));
          }

          suggestionCache.set(key, (result = source.map(toSuggestion)));
        } else {
          if (length < 2) {
            threshold = matchSorter.rankings.CONTAINS;
          }
          const search = spacify(input);

          suggestionCache.set(
            key,
            (result = matchSorter(source, search, {
              keys: ['filter'],
              threshold,
              baseSort: (a, b) => compareSuggestions(a.item, b.item, prefix),
            }).map(toSuggestion)),
          );
        }
      }

      return result;
    },
  };

  function generateClassDescription(className: string): string {
    const css = context.generateCSS(className);

    const bodyStart = css.lastIndexOf('{');
    const bodyEnd = css.indexOf('}');

    if (bodyStart === -1 || bodyEnd === -1) {
      return css;
    }

    return cssbeautify(`.x{${css.slice(bodyStart + 1, bodyEnd)}}`, { indent: '' })
      .split('\n')
      .slice(1, -1)
      .join(' ');
  }

  function toSuggestion(suggestion: Suggestion): Suggestion {
    if (suggestion.type === 'variant') {
      return {
        type: suggestion.type,
        name: suggestion.name,
        value: suggestion.value,
        description: suggestion.description!,
        detail: suggestion.detail!,
        color: suggestion.color!,
      };
    }

    return {
      type: suggestion.type,
      name: suggestion.name,
      value: suggestion.value,
      description: (suggestion.description ||= generateClassDescription(suggestion.name)),
      detail: suggestion.detail!,
      color: suggestion.color!,
    };
  }
}
