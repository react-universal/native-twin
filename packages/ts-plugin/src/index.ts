import { TailwindPluginCreator } from './PluginCreator';

export = function init(modules: {
  typescript: typeof import('typescript/lib/tsserverlibrary');
}) {
  return new TailwindPluginCreator(modules.typescript);
};
