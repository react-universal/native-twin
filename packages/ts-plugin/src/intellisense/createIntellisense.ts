import {
  AutocompleteContext,
  Preset,
  TwindUserConfig,
  defineConfig,
  twind,
  virtual,
} from '@twind/core';
import { CssResolver } from '@universal-labs/css';
import { tw as internalTW } from '@universal-labs/twind-adapter';
import { ConfigurationManager } from '../language-service/configuration';
import type { CurrentTheme, VariantCompletionItem } from '../types';
import { formatCss } from '../utils';
import { createContextExecutor } from './intellisense.compose';

export function createIntellisense() {
  const variants = new Map<string, VariantCompletionItem>();
  // const cssCache = new Map<string, string>();
  const config = defineConfig({
    preflight: false,
    presets: [
      defineConfig<CurrentTheme, Preset<CurrentTheme>[]>({
        ...internalTW.config,
        preflight: false,
      }) as TwindUserConfig<CurrentTheme>,
    ],
    rules: [[ConfigurationManager.VARIANT_MARKER_RULE, { '…': '…' }]],
    ignorelist: [/-\[…]$/],
  }) as TwindUserConfig<CurrentTheme>;
  const tw = twind(config, virtual(false));
  const context: AutocompleteContext<CurrentTheme> = {
    get theme() {
      return tw.theme;
    },
    get variants() {
      return Object.fromEntries(
        Array.from(variants.values(), (variant) => [variant.name.slice(0, -1), variant.name]),
      );
    },
  };
  let nextIndex = 0;
  const ruleExecutor = createContextExecutor(context, tw.theme());
  for (const rule of tw.config.rules) {
    const location = {
      index: nextIndex++,
      position: 0,
    };
    ruleExecutor.run(rule, location);
  }
  // for (const variant of tw.config.variants) {
  //   console.log('VV: ', variant);
  // }
  return {
    getCss,
    classes: ruleExecutor.suggestions,
    context,
  };

  function getCss(name: string) {
    const className = tw(name);
    const target = [...tw.target];
    tw.clear();
    const sheet = CssResolver(target, {
      colorScheme: 'light',
      debug: false,
      deviceHeight: 1080,
      deviceWidth: 720,
      platform: 'ios',
      rem: 16,
    });
    return {
      className,
      sheet,
      css: formatCss(target),
    };
  }
}

export type CreateIntellisenseFn = ReturnType<typeof createIntellisense>;
