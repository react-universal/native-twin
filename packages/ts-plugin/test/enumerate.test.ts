import { test, expect } from 'vitest';

import presetTailwind from '@twind/preset-tailwind';

import { createIntellisense } from '../src/intellisense';

test('enumerate', () => {
  const intellisense = createIntellisense({
    presets: [presetTailwind()],
    theme: {
      data: {
        checked: 'ui~="checked"',
      },
      supports: {
        grid: 'display: grid',
      },
      extend: {
        aria: {
          asc: 'sort="ascending"',
          desc: 'sort="descending"',
        },
      },
    },
  });

  expect(Array.from(intellisense.enumerate(), ({ name }) => name)).toMatchSnapshot();
});
