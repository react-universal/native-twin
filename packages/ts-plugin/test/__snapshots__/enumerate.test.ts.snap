// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`enumerate 1`] = `
[
  "sm:",
  "md:",
  "lg:",
  "xl:",
  "2xl:",
  "active:",
  "after:",
  "any-link:",
  "backdrop:",
  "before:",
  "checked:",
  "contrast-less:",
  "contrast-more:",
  "dark:",
  "default:",
  "defined:",
  "disabled:",
  "empty:",
  "enabled:",
  "even:",
  "file:",
  "first-child:",
  "first-letter:",
  "first-line:",
  "first-of-type:",
  "first:",
  "focus-within:",
  "focus-visible:",
  "focus:",
  "fullscreen:",
  "hover:",
  "in-range:",
  "indeterminate:",
  "invalid:",
  "landscape:",
  "last-child:",
  "last-of-type:",
  "last:",
  "link:",
  "ltr:",
  "marker:",
  "motion-safe:",
  "motion-reduce:",
  "odd:",
  "only-child:",
  "only-of-type:",
  "only:",
  "open:",
  "optional:",
  "out-of-range:",
  "paused:",
  "picture-in-picture:",
  "placeholder-shown:",
  "placeholder:",
  "playing:",
  "portrait:",
  "print:",
  "read-write:",
  "read-only:",
  "required:",
  "rtl:",
  "selection:",
  "sticky:",
  "target:",
  "valid:",
  "visited:",
  "absolute",
  "animate",
  "antialiased",
  "backdrop-sepia",
  "backdrop-saturate",
  "backdrop-invert",
  "backdrop-grayscale",
  "backdrop-contrast",
  "backdrop-brightness",
  "backdrop-blur",
  "backdrop-filter",
  "backdrop-opacity",
  "backdrop-hue-rotate",
  "backdrop-filter-none",
  "-backdrop-hue-rotate",
  "bg-bottom",
  "bg-center",
  "bg-left",
  "bg-right",
  "bg-top",
  "bg-repeat",
  "bg-fixed",
  "bg-local",
  "bg-scroll",
  "bg-left-bottom",
  "bg-left-top",
  "bg-right-bottom",
  "bg-right-top",
  "bg-clip-border",
  "bg-clip-content",
  "bg-clip-padding",
  "bg-clip-text",
  "bg-no-repeat",
  "bg-repeat-round",
  "bg-repeat-space",
  "bg-repeat-x",
  "bg-repeat-y",
  "bg-origin-border",
  "bg-origin-content",
  "bg-origin-padding",
  "bg-gradient-to-b",
  "bg-gradient-to-bl",
  "bg-gradient-to-br",
  "bg-gradient-to-l",
  "bg-gradient-to-r",
  "bg-gradient-to-t",
  "bg-gradient-to-tl",
  "bg-gradient-to-tr",
  "block",
  "blur",
  "border",
  "border-b",
  "border-l",
  "border-r",
  "border-t",
  "border-x",
  "border-y",
  "border-spacing",
  "border-dashed",
  "border-dotted",
  "border-double",
  "border-none",
  "border-solid",
  "border-collapse",
  "border-separate",
  "border-opacity",
  "border-spacing-x",
  "border-spacing-y",
  "bottom",
  "-bottom",
  "box-border",
  "box-content",
  "box-decoration-clone",
  "box-decoration-slice",
  "break-keep",
  "break-all",
  "break-words",
  "break-normal",
  "brightness",
  "capitalize",
  "clear-both",
  "clear-left",
  "clear-none",
  "clear-right",
  "col-end-auto",
  "col-start-auto",
  "collapse",
  "container",
  "contents",
  "contrast",
  "decoration-clone",
  "decoration-slice",
  "delay",
  "diagonal-fractions",
  "divide-x",
  "divide-y",
  "divide-dashed",
  "divide-dotted",
  "divide-double",
  "divide-none",
  "divide-solid",
  "divide-opacity",
  "divide-x-reverse",
  "divide-y-reverse",
  "drop-shadow",
  "duration",
  "ease",
  "filter",
  "filter-none",
  "fixed",
  "flex",
  "flex-grow",
  "flex-shrink",
  "flex-nowrap",
  "flex-wrap",
  "flex-col",
  "flex-row",
  "flex-wrap-reverse",
  "flex-col-reverse",
  "flex-row-reverse",
  "float-left",
  "float-none",
  "float-right",
  "flow-root",
  "gap",
  "gap-y",
  "gap-x",
  "grayscale",
  "grid",
  "grid-flow-dense",
  "grid-flow-col",
  "grid-flow-row",
  "grid-flow-col-dense",
  "grid-flow-row-dense",
  "grow",
  "hidden",
  "hue-rotate",
  "-hue-rotate",
  "inline",
  "inline-block",
  "inline-flex",
  "inline-grid",
  "inline-table",
  "inset",
  "inset-x",
  "inset-y",
  "-inset",
  "-inset-x",
  "-inset-y",
  "invert",
  "invisible",
  "isolate",
  "isolation-auto",
  "italic",
  "left",
  "-left",
  "line-through",
  "lining-nums",
  "list-inside",
  "list-outside",
  "list-item",
  "lowercase",
  "m",
  "-m",
  "mb",
  "-mb",
  "ml",
  "-ml",
  "mr",
  "-mr",
  "mt",
  "-mt",
  "mx",
  "-mx",
  "my",
  "-my",
  "no-underline",
  "normal-case",
  "normal-nums",
  "not-italic",
  "not-sr-only",
  "object-bottom",
  "object-center",
  "object-left",
  "object-right",
  "object-top",
  "object-contain",
  "object-cover",
  "object-fill",
  "object-none",
  "object-left-bottom",
  "object-left-top",
  "object-right-bottom",
  "object-right-top",
  "object-scale-down",
  "oldstyle-nums",
  "ordinal",
  "origin-bottom",
  "origin-center",
  "origin-left",
  "origin-right",
  "origin-top",
  "origin-bottom-left",
  "origin-bottom-right",
  "origin-top-left",
  "origin-top-right",
  "outline",
  "outline-dashed",
  "outline-dotted",
  "outline-double",
  "outline-none",
  "overflow-auto",
  "overflow-clip",
  "overflow-hidden",
  "overflow-scroll",
  "overflow-visible",
  "overflow-x-auto",
  "overflow-x-clip",
  "overflow-x-hidden",
  "overflow-x-scroll",
  "overflow-x-visible",
  "overflow-y-auto",
  "overflow-y-clip",
  "overflow-y-hidden",
  "overflow-y-scroll",
  "overflow-y-visible",
  "overline",
  "overscroll-auto",
  "overscroll-contain",
  "overscroll-none",
  "overscroll-x-auto",
  "overscroll-x-contain",
  "overscroll-x-none",
  "overscroll-y-auto",
  "overscroll-y-contain",
  "overscroll-y-none",
  "p",
  "pb",
  "pl",
  "pr",
  "proportional-nums",
  "pt",
  "px",
  "py",
  "relative",
  "resize",
  "resize-none",
  "resize-x",
  "resize-y",
  "right",
  "-right",
  "ring",
  "ring-inset",
  "ring-offset",
  "ring-opacity",
  "rounded",
  "rounded-b",
  "rounded-bl",
  "rounded-br",
  "rounded-l",
  "rounded-r",
  "rounded-t",
  "rounded-tl",
  "rounded-tr",
  "row-end-auto",
  "row-start-auto",
  "saturate",
  "scroll-m",
  "scroll-mb",
  "scroll-ml",
  "scroll-mr",
  "scroll-mt",
  "scroll-mx",
  "scroll-my",
  "scroll-p",
  "scroll-pb",
  "scroll-pl",
  "scroll-pr",
  "scroll-pt",
  "scroll-px",
  "scroll-py",
  "scroll-auto",
  "scroll-smooth",
  "-scroll-m",
  "-scroll-mb",
  "-scroll-ml",
  "-scroll-mr",
  "-scroll-mt",
  "-scroll-mx",
  "-scroll-my",
  "select-all",
  "select-auto",
  "select-none",
  "select-text",
  "sepia",
  "shadow",
  "shrink",
  "slashed-zero",
  "snap-always",
  "snap-normal",
  "snap-center",
  "snap-end",
  "snap-start",
  "snap-mandatory",
  "snap-proximity",
  "snap-both",
  "snap-x",
  "snap-y",
  "snap-none",
  "snap-align-none",
  "space-x",
  "space-y",
  "space-x-reverse",
  "space-y-reverse",
  "-space-x",
  "-space-y",
  "sr-only",
  "stacked-fractions",
  "static",
  "sticky",
  "subpixel-antialiased",
  "table",
  "table-caption",
  "table-cell",
  "table-column",
  "table-row",
  "table-auto",
  "table-fixed",
  "table-column-group",
  "table-footer-group",
  "table-header-group",
  "table-row-group",
  "tabular-nums",
  "text-clip",
  "text-ellipsis",
  "text-center",
  "text-end",
  "text-justify",
  "text-left",
  "text-right",
  "text-start",
  "top",
  "-top",
  "touch-auto",
  "touch-manipulation",
  "touch-none",
  "touch-pan-down",
  "touch-pan-left",
  "touch-pan-right",
  "touch-pan-up",
  "touch-pan-x",
  "touch-pan-y",
  "touch-pinch-zoom",
  "transform",
  "transform-cpu",
  "transform-gpu",
  "transform-none",
  "transition",
  "truncate",
  "underline",
  "uppercase",
  "visible",
]
`;
